### DISCUSSION TITLE
Export data using a set of coordinates

### DISCUSSION THREAD

**Discussion by nicolorossetti** [MPh-py/MPh#101] [General]
Hi everyone,

First of all, thanks for MPh, it's really a great library!
I have a question regarding the extraction of solution results using a set of coordinates.

In my case, I am currently evaluating the electric potential in space using: (x, y, z, V) = model.evaluate(['x', 'y', 'z', 'V'])
Is there a way to specify a set of x,y,z coordinates where to evaluate the electric potential?
In the COMSOL GUI there is the option to select a grid of points, for example setting x as range(0,20,1000). Is it possible to do the same with MPh?

Thank you in advance!

Nicolo

**Comment by john-hen** [comment]
Hi Nicolo,
I'm not quite sure what you mean, like what exactly you're doing in the GUI. If you're using a grid dataset (dataset type "Grid 2D" or "Grid 3d"), then that is not supported by `model.evaluate()`, see remarks in issue #26. These remarks refer to MPh's high-level API. If you use the low-level Java calls (as explained in [chapter "Demonstrations"](https://mph.readthedocs.io/en/stable/demonstrations.html#access-the-full-comsol-api)), then it is of course possible: everything you can do in Java, you can also do in Python. But then the answer would be (somewhere) in the [Comsol documentation](https://comsol.com/documentation/COMSOL_ProgrammingReferenceManual.pdf).

**Reply by nicolorossetti** [reply]
Hi John,

Thank you for your quick reply!
I am referring to the options that are present in the Comsol GUI when exporting data (see image).

![Picture1](https://user-images.githubusercontent.com/86466814/200251484-f8b25f7c-c156-49ff-904d-1eda7f01cede.png)

When on python I use (x, y, z, V) = model.evaluate(['x', 'y', 'z', 'V']), the electric potential is evaluated in a non-uniform grid of points. This in the GUI is equivalent to select the "Select from Dataset" option. What I would like to do is to evaluate the electric potential in a uniform (equally spaced) grid of points along all axes. I am still quite new to Comsol, but from what I understood my dataset is directly mapping to the solution.

Is this supported by the high-level API of MPh?

**Comment by john-hen** [comment]
Hi Nicolo,

According to your reply, you are modifying the [properties](https://mph.readthedocs.io/en/1.2/api/mph.Node.html#mph.Node.properties) of an export node, in your case called `Data 1`.

Here's a code example that does something similar with the model from the [`demos` folder](https://github.com/MPh-py/MPh/tree/main/demos):
```python
import mph

# Load and solve the model.
client = mph.start()
model = client.load('capacitor.mph')
model.solve('static')

# Export data on structured grid.
node = model/'exports'/'data'
node.property("gridstruct", "spreadsheet")
node.property("gridx2", "range(-0.01, 0.001, +0.01)")
node.property("gridy2", "range(-0.01, 0.001, +0.01)")
model.export(node, 'data.txt')
```

Here the export node is called `data`, and is defined in the model itself. I just changed some of its properties. The node could also be created from scratch if need be, but usually I start with a base model, and then just modify whatever needs scripting.

What all these properties are called would be explained in the Comsol documentation. This is not replicated in the MPh documentation, as there are hundreds if not thousands of properties, across all node types. So what I do in such a case, I go through the steps in the GUI, save the model as a `.java` file, look at the last few lines of code, and put that back into Python syntax. (The Java-like syntax would also work, with minor modifications, as explained in the documentation).

**Reply by nicolorossetti** [reply]
Hi John,

Thanks for the info, it was very useful.
I have tried your steps but they seem not to be working in my case, probably because the node is not there yet but it gets created when I use the GUI (i.e. I am not modifying the node but I am creating it. I don't want to create the node manually on the GUI every time but I would like to do it through the code). I have saved the model as a .java file to have a look at the commands and I get the following last lines:

```
    model.result().create("pg1", "PlotGroup3D");
    model.result("pg1").label("Electric Potential (ec)");
    model.result("pg1").set("frametype", "spatial");
    model.result("pg1").set("data", "dset1");
    model.result("pg1").feature().create("mslc1", "Multislice");
    model.result("pg1").feature("mslc1").set("colortable", "RainbowLight");
    model.result("pg1").feature("mslc1").set("data", "parent");
    model.result("pg1").run();
    model.result("pg1").set("data", "dset1");

    model.component("comp1").physics("ec").feature("pcs1").set("Qjp", 0.001);

    model.sol("sol1").runAll();

    model.label("0.mph");

    model.result().export().create("data1", "Data");
    model.result().export("data1").set("expr", new String[]{"V"});
    model.result().export("data1").set("descr", new String[]{"Electric potential"});
    model.result().export("data1").set("filename", "test");
    model.result().export("data1").set("location", "grid");
    model.result().export("data1").set("gridx3", "range(-1000,20,1000)");
    model.result().export("data1").set("gridy3", "range(-1000,20,1000)");
    model.result().export("data1").set("gridz3", "range(5000,50,10000)");
    model.result().export("data1").run();
```

following these code lines, I have tried to run the following lines to create the node first:

```
node = model/'result'/'exports'
node.create("data1" ,"Data")
```

However, I get the following error:

Node plots/exports does not support feature creation.
Traceback (most recent call last):
  File "/venv/lib/python3.9/site-packages/IPython/core/interactiveshell.py", line 3444, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-e7b6e6317e19>", line 1, in <module>
    runfile('/export_electric_fields.py', wdir='/src_analysis')
  File "/pycharm-community-2021.3.2/plugins/python-ce/helpers/pydev/_pydev_bundle/pydev_umd.py", line 198, in runfile
    pydev_imports.execfile(filename, global_vars, local_vars)  # execute the script
  File "/pycharm-community-2021.3.2/plugins/python-ce/helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/export_electric_fields.py", line 13, in <module>
    node.create("data1" ,"Data")
  File "/venv/lib/python3.9/site-packages/mph/node.py", line 597, in create
    raise RuntimeError(error)
RuntimeError: Node plots/exports does not support feature creation.

I did not manage to understand so far what is the problem.

**Reply by john-hen** [reply]
I think the syntax to create the node should be:
```python
node = model/'exports'
node.create("Data", name="data")
```
Haven't tried it, this is just based on what I used in [`create_capacitor.py`](https://github.com/MPh-py/MPh/blob/main/demos/create_capacitor.py).

**Reply by nicolorossetti** [reply]
Thanks John,

I finally managed to make it work! 
Just as a reference, here is the final code:

```
# Export data on structured grid.
exportnode = model/'exports'
exportnode.create("Data", name="data")

datanode = exportnode/"data"
datanode.property("expr", ["V"])
datanode.property("descr", ["Electric potential"])
datanode.property("location", "grid")
datanode.property("gridstruct", "spreadsheet")
datanode.property("gridx3", "range(-1000, 20, +1000)")
datanode.property("gridy3", "range(-1000, 20, +1000)")
datanode.property("gridz3", "range(+5000, 50, +10000)")

model.export(datanode, 'data_pymph.csv')
```
